pkg/virtual/workspaces/registry/rest.go

```diff
@@ -0,0 +1,547 @@
+			Name:     adminRoleBindingName,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "User",
+				Name:      user.GetName(),
+				Namespace: "",
+			},
+		},
+	}
+	if _, err := s.rbacClient.ClusterRoleBindings().Create(ctx, &clusterRoleBinding, metav1.CreateOptions{}); err != nil {
+		if kerrors.IsAlreadyExists(err) {
+			return nil, kerrors.NewAlreadyExists(tenancyv1alpha1.Resource("workspaces"), workspace.Name)
+		}
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+	}
+
+	// Then create the admin and lister roles related to the given workspace.
+	// Note that ResourceNames contains the workspace pretty name for now.
+	// It will be updated later on when the internal name of the workspace is known.
+	adminClusterRole := rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   adminRoleBindingName,
+			Labels: map[string]string{},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:         roleRules[AdminRoleType].Verbs,
+				APIGroups:     []string{tenancyv1alpha1.SchemeGroupVersion.Group},
+				Resources:     roleRules[AdminRoleType].Resources,
+				ResourceNames: []string{workspace.Name},
+			},
+		},
+	}
+	if _, err := s.rbacClient.ClusterRoles().Create(ctx, &adminClusterRole, metav1.CreateOptions{}); err != nil && !kerrors.IsAlreadyExists(err) {
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+	}
+
+	listerClusterRole := rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   listerRoleBindingName,
+			Labels: map[string]string{},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:         roleRules[ListerRoleType].Verbs,
+				APIGroups:     []string{tenancyv1alpha1.SchemeGroupVersion.Group},
+				Resources:     roleRules[ListerRoleType].Resources,
+				ResourceNames: []string{workspace.Name},
+			},
+		},
+	}
+	if _, err := s.rbacClient.ClusterRoles().Create(ctx, &listerClusterRole, metav1.CreateOptions{}); err != nil && !kerrors.IsAlreadyExists(err) {
+		_ = s.rbacClient.ClusterRoles().Delete(ctx, adminClusterRole.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+	}
+
+	// Then try to create the workspace object itself, first with the pretty name,
+	// retrying with increasing suffixes until a workspace with the same name
+	// doesn't already exist.
+	// The suffixed name based on the pretty name will be the internal name
+	prettyName := workspace.Name
+	var createdWorkspace *tenancyv1alpha1.Workspace
+	var err error
+	var nameSuffix string
+	i := 0
+	for i < 10 {
+		if i > 0 {
+			nameSuffix = fmt.Sprintf("-%d", i)
+			workspace.Name = fmt.Sprintf("%s-%s", prettyName, nameSuffix)
+		}
+		createdWorkspace, err = s.workspaceClient.Create(ctx, workspace, metav1.CreateOptions{})
+		if err == nil {
+			break
+		}
+		if !kerrors.IsAlreadyExists(err) {
+			return nil, err
+		}
+		i++
+	}
```

Comment:
how do you know it has succeeded here?

Reply:
Good catch. I'm missing an error test after the loop

```go
	if err != nil {
		return nil, err
	}
```
Let me add it

Reply:
Fixed in commit ee865a2f2aeef887da5232a5bfc0200dde8ca3b4